wen-restart/src/wen_restart.rs

```diff
@@ -480,4 +521,116 @@ mod tests {
+        let wen_restart_repair_slots = Some(Arc::new(RwLock::new(Vec::new())));
+        let mut last_vote_fork_slots = Vec::new();
+        let (_, blockstore, cluster_info, bank_forks) = wen_restart_test_init(
+            &ledger_path,
+            expected_slots,
+            shred_version,
+            &mut last_vote_fork_slots,
+        );
+        let init_progress = WenRestartProgress {
+            state: RestartState::Init.into(),
+            my_last_voted_fork_slots: None,
+            last_voted_fork_slots_aggregate: None,
+        };
+        let last_vote_slot: Slot = (RestartLastVotedForkSlots::MAX_SLOTS + expected_slots)
+            .try_into()
+            .unwrap();
+        let last_vote_bankhash = Hash::new_unique();
+        let last_voted_fork_slots_progress = WenRestartProgress {
+            state: RestartState::LastVotedForkSlots.into(),
+            my_last_voted_fork_slots: Some(LastVotedForkSlotsRecord {
+                last_vote_fork_slots: vec![last_vote_slot],
+                last_vote_bankhash: last_vote_bankhash.to_string(),
+                shred_version: shred_version as u32,
+            }),
+            last_voted_fork_slots_aggregate: None,
+        };
+        for (progress, last_vote, setup, expected_err, cleanup) in vec![
+            (
+                init_progress.clone(),
+                VoteTransaction::from(VoteStateUpdate::from(vec![(0, 8), (last_vote_slot, 1)])),
+                None,
+                "last vote is not a vote transaction",
+                None,
+            ),
+            (
+                init_progress,
+                VoteTransaction::from(Vote::new(vec![], Hash::new_unique())),
+                None,
+                "last vote does not have last voted slot",
+                None,
+            ),
+            (
+                last_voted_fork_slots_progress,
+                VoteTransaction::from(Vote::new(vec![last_vote_slot], last_vote_bankhash)),
+                Some(|| change_proto_file_readonly(&wen_restart_proto_path, true)),
+                "Permission denied",
+                Some(|| change_proto_file_readonly(&wen_restart_proto_path, false)),
+            ),
+        ] {
+            assert!(write_wen_restart_records(&wen_restart_proto_path, &progress,).is_ok());
+            if let Some(setup_function) = setup {
+                setup_function();
+            }
+            assert_matches!(wait_for_wen_restart(
+                &wen_restart_proto_path,
+                last_vote,
+                blockstore.clone(),
+                cluster_info.clone(),
+                bank_forks.clone(),
+                wen_restart_repair_slots.clone(),
+                80,
+                Arc::new(AtomicBool::new(false)),
+            ).err().unwrap().to_string(), message if message.contains(expected_err));
+            if let Some(cleanup_function) = cleanup {
+                cleanup_function();
+            }
```

Comment:
Nice, I think this is a step in the right direction!

However, this doesn't seem to be testing that we can actually recover from failure. 

I think we want to show that on failure, we can:

1. stored the saved state
2. read from the saved state
3. restart, call `wait_for_wen_restart()` and still successfully move onto the next state transition.

Reply:
To this end, I think it will be worth separating each test by `RestartState`

Reply:
Reorganized tests a bit, does this look better?